### PR TITLE
Make the JIT use a single consistent vmentry for each MethodHandle

### LINKED ISSUE #13588: MathLoadTest_bigdecimal_special_5m crash in compiled method vmState=0x00000000
https://openj9-jenkins.osuosl.org/job/Test_openjdk17_j9_special.system_ppc64le_linux_Release_testList_1/6
MathLoadTest_bigdecimal_special_5m_9
-Xaggressive -Xgcpolicy:gencon -Xjit -Xnocompressedrefs

https://openj9-artifactory.osuosl.org/artifactory/ci-openj9/Test/Test_openjdk17_j9_special.system_ppc64le_linux_Release_testList_1/6/system_test_output.tar.gz
```
MLT stderr Unhandled exception
MLT stderr Type=Segmentation error vmState=0x00000000
MLT stderr Unhandled exception
MLT stderr Type=Segmentation error vmState=0x00000000
MLT stderr J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
MLT stderr Handler1=00003FFF861C4970 Handler2=00003FFF860FC300
MLT stderr J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
MLT stderr Handler1=00003FFF861C4970 Handler2=00003FFF860FC300
MLT stderr R0=00003FFEE5B11428 R1=00003FFEE537B740 R2=00000000000000EC R3=000000010000006A
MLT stderr R4=00003FFF7F54BE30 R5=00003FFF7F54BE10 R6=0000000000000000 R7=0000000000000000
MLT stderr R8=00003FFF007D92B0 R9=00003FFF7F54BF1C R10=00003FFEE5B11360 R11=00003FFF850BDC78
MLT stderr R12=0000000048004822 R13=00003FFEE5386900 R14=00003FFE40008130 R15=00003FFF80A93200
MLT stderr R16=00003FFEC7D10038 R17=FFFFFFFFFFFFFFFF R18=0000000000000000 R19=0000000000000000
MLT stderr R20=0000000000000000 R21=00003FFF010ED030 R22=00003FFF011363D8 R23=00003FFF010ED070
MLT stderr R24=00003FFF7F54BE10 R25=00003FFF7F54BE10 R26=00003FFF00A384F0 R27=00003FFF011363D8
MLT stderr R28=00003FFF007D9278 R29=0000000000000000 R30=00003FFF010ED070 R31=00003FFF7F54BE30
MLT stderr R0=00003FFEE5B11428 R1=00003FFEE517B740 R2=00000000000000EC R3=000000010000006A
MLT stderr R4=00003FFF7F5075A0 R5=00003FFF7F507580 R6=0000000000000000 R7=0000000000000000
MLT stderr R8=00003FFF007D92B0 R9=00003FFF7F50768C R10=00003FFEE5B11360 R11=00003FFF850BDC78
MLT stderr R12=0000000048004822 R13=00003FFEE5186900 R14=00003FFE30008130 R15=00003FFF80AB0300
MLT stderr R16=00003FFEC7D10038 R17=FFFFFFFFFFFFFFFF R18=0000000000000000 R19=0000000000000000
MLT stderr R20=0000000000000000 R21=00003FFF00FB8C08 R22=00003FFF011363D8 R23=00003FFF00FB8C48
MLT stderr R24=00003FFF7F507580 R25=00003FFF7F507580 R26=00003FFF00A38290 R27=00003FFF011363D8
MLT stderr R28=00003FFF007D9278 R29=0000000000000000 R30=00003FFF00FB8C48 R31=00003FFF7F5075A0
MLT stderr NIP=00003FFEE5B11434 MSR=800000010280F033 ORIG_GPR3=00000000000081C8 CTR=00003FFF850BDC78
MLT stderr LINK=00003FFEE5B11428 XER=0000000000000000 CCR=0000000028004824 SOFTE=0000000000000001
MLT stderr TRAP=0000000000000300 DAR=0000000000000050 dsisr=0000000040000000 RESULT=0000000000000000
MLT stderr NIP=00003FFEE5B11434 MSR=800000010280F033 ORIG_GPR3=00000000000081C8 CTR=00003FFF850BDC78
MLT stderr LINK=00003FFEE5B11428 XER=0000000000000000 CCR=0000000028004824 SOFTE=0000000000000001
MLT stderr TRAP=0000000000000300 DAR=0000000000000050 dsisr=0000000040000000 RESULT=0000000000000000
MLT stderr Unhandled exception
MLT stderr Type=Segmentation error vmState=0x00000000
MLT stderr Unhandled exception
MLT stderr Type=Segmentation error vmState=0x00000000
MLT stderr Unhandled exception
MLT stderr Type=Segmentation error vmState=0x00000000
MLT stderr J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
MLT stderr Handler1=00003FFF861C4970 Handler2=00003FFF860FC300
MLT stderr J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
MLT stderr Handler1=00003FFF861C4970 Handler2=00003FFF860FC300
MLT stderr J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
MLT stderr Handler1=00003FFF861C4970 Handler2=00003FFF860FC300
MLT stderr Unhandled exception
MLT stderr Type=Segmentation error vmState=0x00000000
MLT stderr R0=00003FFEE5B11428 R1=00003FFEE52FB740 R2=00000000000000EC R3=000000010000006A
MLT stderr R4=00003FFF7F55EB68 R5=00003FFF7F55EB48 R6=0000000000000000 R7=0000000000000000
MLT stderr R8=00003FFF007D92B0 R9=00003FFF7F55EC54 R10=00003FFEE5B11360 R11=00003FFF850BDC78
MLT stderr R12=0000000048004822 R13=00003FFEE5306900 R14=00003FFE34008090 R15=00003FFF80A9AB00
MLT stderr R16=00003FFEC7D10038 R17=FFFFFFFFFFFFFFFF R18=0000000000000000 R19=0000000000000000
MLT stderr R20=0000000000000000 R21=00003FFF0109D360 R22=00003FFF011363D8 R23=00003FFF0109D3A0
MLT stderr R24=00003FFF7F55EB48 R25=00003FFF7F55EB48 R26=00003FFF00A38160 R27=00003FFF011363D8
MLT stderr R28=00003FFF007D9278 R29=0000000000000000 R30=00003FFF0109D3A0 R31=00003FFF7F55EB68
MLT stderr J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
MLT stderr Handler1=00003FFF861C4970 Handler2=00003FFF860FC300
MLT stderr R0=00003FFEE5B11428 R1=00003FFEE507B740 R2=00000000000000EC R3=000000010000006A
MLT stderr R4=00003FFF7F471828 R5=00003FFF7F471808 R6=0000000000000000 R7=0000000000000000
MLT stderr R8=00003FFF007D92B0 R9=00003FFF7F471914 R10=00003FFEE5B11360 R11=00003FFF850BDC78
MLT stderr R12=0000000048004822 R13=00003FFEE5086900 R14=00003FFE28008130 R15=00003FFF80ABD200
MLT stderr R16=00003FFEC7D10038 R17=FFFFFFFFFFFFFFFF R18=0000000000000000 R19=0000000000000000
MLT stderr R20=0000000000000000 R21=00003FFF010C94E0 R22=00003FFF011363D8 R23=00003FFF010C9520
MLT stderr R24=00003FFF7F471808 R25=00003FFF7F471808 R26=00003FFF00A38328 R27=00003FFF011363D8
MLT stderr R28=00003FFF007D9278 R29=0000000000000000 R30=00003FFF010C9520 R31=00003FFF7F471828
MLT stderr NIP=00003FFEE5B11434 MSR=800000010280F033 ORIG_GPR3=00000000000081C8 CTR=00003FFF850BDC78
MLT stderr LINK=00003FFEE5B11428 XER=0000000000000000 CCR=0000000028004824 SOFTE=0000000000000001
MLT stderr TRAP=0000000000000300 DAR=0000000000000050 dsisr=0000000040000000 RESULT=0000000000000000
MLT stderr R0=00003FFEE5B11428 R1=00003FFEE547B740 R2=00000000000000EC R3=000000010000006A
MLT stderr R4=00003FFF7F539CD0 R5=00003FFF7F539CB0 R6=0000000000000000 R7=0000000000000000
MLT stderr R8=00003FFF007D92B0 R9=00003FFF7F539DBC R10=00003FFEE5B11360 R11=00003FFF850BDC78
MLT stderr R12=0000000048004822 R13=00003FFEE5486900 R14=00003FFE48009180 R15=00003FFF80A85900
MLT stderr R16=00003FFEC7D10038 R17=FFFFFFFFFFFFFFFF R18=0000000000000000 R19=0000000000000000
MLT stderr R20=0000000000000000 R21=00003FFF00F46F58 R22=00003FFF011363D8 R23=00003FFF00F46F98
MLT stderr R24=00003FFF7F539CB0 R25=00003FFF7F539CB0 R26=00003FFF00A65618 R27=00003FFF011363D8
MLT stderr R28=00003FFF007D9278 R29=0000000000000000 R30=00003FFF00F46F98 R31=00003FFF7F539CD0
MLT stderr NIP=00003FFEE5B11434 MSR=800000010280F033 ORIG_GPR3=00000000000081C8 CTR=00003FFF850BDC78
MLT stderr LINK=00003FFEE5B11428 XER=0000000000000000 CCR=0000000028004824 SOFTE=0000000000000001
MLT stderr TRAP=0000000000000300 DAR=0000000000000050 dsisr=0000000040000000 RESULT=0000000000000000
MLT stderr NIP=00003FFEE5B11434 MSR=800000010280F033 ORIG_GPR3=00000000000081C8 CTR=00003FFF850BDC78
MLT stderr LINK=00003FFEE5B11428 XER=0000000000000000 CCR=0000000028004824 SOFTE=0000000000000001
MLT stderr TRAP=0000000000000300 DAR=0000000000000050 dsisr=0000000040000000 RESULT=0000000000000000
MLT stderr FPR0 00003ffe40008058 (f: 1073774720.000000, d: 3.476307e-310)
MLT stderr FPR1 4055aae840000000 (f: 1073741824.000000, d: 8.667043e+01)
MLT stderr FPR2 3ff0000000000000 (f: 0.000000, d: 1.000000e+00)
MLT stderr FPR3 bf3d4a1000000000 (f: 0.000000, d: -4.469194e-04)
MLT stderr FPR4 3f499a0000000000 (f: 0.000000, d: 7.812977e-04)
MLT stderr FPR5 bf8b0c0d60000000 (f: 1610612736.000000, d: -1.320658e-02)
MLT stderr FPR6 3f74257300000000 (f: 0.000000, d: 4.918527e-03)
MLT stderr FPR7 3ef95df720000000 (f: 536870912.000000, d: 2.419191e-05)
MLT stderr FPR8 0074007300650074 (f: 6619252.000000, d: 1.780215e-306)
MLT stderr FPR9 006d006500740049 (f: 7602249.000000, d: 1.290611e-306)
MLT stderr FPR10 003100330030005f (f: 3145823.000000, d: 9.456997e-308)
MLT stderr FPR11 37f0001d00000000 (f: 0.000000, d: 2.938817e-39)
MLT stderr FPR12 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR13 bfb7440000000000 (f: 0.000000, d: -9.088135e-02)
MLT stderr FPR14 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR15 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR16 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR17 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR18 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR19 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR20 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR21 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR22 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR23 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR24 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR25 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR26 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR27 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR28 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR29 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR30 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR31 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr R0=00003FFEE5B11428 R1=00003FFEE4F7B740 R2=00000000000000EC R3=000000010000006A
MLT stderr R4=00003FFF7F4F15D0 R5=00003FFF7F4F15B0 R6=0000000000000000 R7=0000000000000000
MLT stderr R8=00003FFF007D92B0 R9=00003FFF7F4F16BC R10=00003FFEE5B11360 R11=00003FFF850BDC78
MLT stderr R12=0000000048004822 R13=00003FFEE4F86900 R14=00003FFE20008130 R15=00003FFF80ACAB00
MLT stderr R16=00003FFEC7D10038 R17=FFFFFFFFFFFFFFFF R18=0000000000000000 R19=0000000000000000
MLT stderr R20=0000000000000000 R21=00003FFF01094198 R22=00003FFF011363D8 R23=00003FFF010941D8
MLT stderr R24=00003FFF7F4F15B0 R25=00003FFF7F4F15B0 R26=00003FFF00A17C90 R27=00003FFF011363D8
MLT stderr R28=00003FFF007D9278 R29=0000000000000000 R30=00003FFF010941D8 R31=00003FFF7F4F15D0
MLT stderr 
MLT stderr Compiled_method=net/adoptopenjdk/loadTest/adaptors/JUnitAdaptor$1.addResult(Ljava/lang/String;Ljava/lang/String;Lorg/junit/runner/Description;Ljava/lang/String;Ljava/lang/Throwable;)V
MLT stderr FPR0 00003ffe30008058 (f: 805339200.000000, d: 3.476293e-310)
MLT stderr FPR1 404e5993c0000000 (f: 3221225472.000000, d: 6.069982e+01)
MLT stderr FPR2 3ff0000000000000 (f: 0.000000, d: 1.000000e+00)
MLT stderr FPR3 3fc7466496cb03de (f: 2529887232.000000, d: 1.818357e-01)
MLT stderr FPR4 3fcc71c51d8e78af (f: 495876256.000000, d: 2.222220e-01)
MLT stderr FPR5 3fd2492494229359 (f: 2485293824.000000, d: 2.857143e-01)
MLT stderr FPR6 3f220b0240000000 (f: 1073741824.000000, d: 1.376572e-04)
MLT stderr FPR7 3e5458cca0000000 (f: 2684354560.000000, d: 1.894950e-08)
MLT stderr FPR8 0074007300650074 (f: 6619252.000000, d: 1.780215e-306)
MLT stderr FPR9 006d006500740049 (f: 7602249.000000, d: 1.290611e-306)
MLT stderr FPR10 003500310030005f (f: 3145823.000000, d: 1.168205e-307)
MLT stderr FPR11 37f0001d00000000 (f: 0.000000, d: 2.938817e-39)
MLT stderr FPR12 4144938b80000000 (f: 2147483648.000000, d: 2.696983e+06)
MLT stderr FPR13 bfb7440000000000 (f: 0.000000, d: -9.088135e-02)
MLT stderr FPR14 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR15 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR16 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR17 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR18 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR19 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR20 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR21 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR22 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR23 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR24 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR25 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR26 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR27 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR28 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR29 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR30 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR31 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr Target=2_90_20210925_6 (Linux 4.4.0-173-generic)
MLT stderr 
MLT stderr Compiled_method=net/adoptopenjdk/loadTest/adaptors/JUnitAdaptor$1.addResult(Ljava/lang/String;Ljava/lang/String;Lorg/junit/runner/Description;Ljava/lang/String;Ljava/lang/Throwable;)V
MLT stderr CPU=ppc64le (16 logical CPUs) (0x1fe090000 RAM)
MLT stderr NIP=00003FFEE5B11434 MSR=800000010280F033 ORIG_GPR3=00000000000081C8 CTR=00003FFF850BDC78
MLT stderr LINK=00003FFEE5B11428 XER=0000000000000000 CCR=0000000028004824 SOFTE=0000000000000001
MLT stderr TRAP=0000000000000300 DAR=0000000000000050 dsisr=0000000040000000 RESULT=0000000000000000
MLT stderr Target=2_90_20210925_6 (Linux 4.4.0-173-generic)
MLT stderr ----------- Stack Backtrace -----------
MLT stderr CPU=ppc64le (16 logical CPUs) (0x1fe090000 RAM)
MLT stderr ----------- Stack Backtrace -----------
MLT stderr Unhandled exception
MLT stderr Type=Segmentation error vmState=0x00000000
MLT stderr J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
MLT stderr Handler1=00003FFF861C4970 Handler2=00003FFF860FC300
MLT stderr FPR0 00003ffe34007fb8 (f: 872447936.000000, d: 3.476297e-310)
MLT stderr FPR1 4055a65740000000 (f: 1073741824.000000, d: 8.659908e+01)
MLT stderr FPR2 3ff0000000000000 (f: 0.000000, d: 1.000000e+00)
MLT stderr FPR3 bf3d4a1000000000 (f: 0.000000, d: -4.469194e-04)
MLT stderr FPR4 3f499a0000000000 (f: 0.000000, d: 7.812977e-04)
MLT stderr FPR5 bf8b0c0d60000000 (f: 1610612736.000000, d: -1.320658e-02)
MLT stderr FPR6 3f7808d4e0000000 (f: 3758096384.000000, d: 5.867797e-03)
MLT stderr FPR7 3f020d41c0000000 (f: 3221225472.000000, d: 3.443105e-05)
MLT stderr FPR8 0074007300650074 (f: 6619252.000000, d: 1.780215e-306)
MLT stderr FPR9 006d006500740049 (f: 7602249.000000, d: 1.290611e-306)
MLT stderr FPR10 003200330030005f (f: 3145823.000000, d: 1.001327e-307)
MLT stderr FPR11 37f0001d00000000 (f: 0.000000, d: 2.938817e-39)
MLT stderr FPR12 4144938b80000000 (f: 2147483648.000000, d: 2.696983e+06)
MLT stderr FPR13 bfb7440000000000 (f: 0.000000, d: -9.088135e-02)
MLT stderr FPR14 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR15 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR16 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR17 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR18 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR19 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR20 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR21 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR22 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR23 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR24 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR25 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR26 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR27 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR28 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR29 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR30 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr FPR31 0000000000000000 (f: 0.000000, d: 0.000000e+00)
MLT stderr R0=00003FFEE5B11428 R1=00003FFEE4EFB740 R2=00000000000000EC R3=000000010000006A
MLT stderr R4=00003FFF7F5281D0 R5=00003FFF7F5281B0 R6=0000000000000000 R7=0000000000000000
MLT stderr R8=00003FFF007D92B0 R9=00003FFF7F5282BC R10=00003FFEE5B11360 R11=00003FFF850BDC78
MLT stderr R12=0000000048004822 R13=00003FFEE4F06900 R14=00003FFE14008130 R15=00003FFF80AD1300
MLT stderr R16=00003FFEC7D10038 R17=FFFFFFFFFFFFFFFF R18=0000000000000000 R19=0000000000000000
MLT stderr R20=0000000000000000 R21=00003FFF010DB098 R22=00003FFF011363D8 R23=00003FFF010DB0D8
MLT stderr R24=00003FFF7F5281B0 R25=00003FFF7F5281B0 R26=00003FFF00A38008 R27=00003FFF011363D8
MLT stderr R28=00003FFF007D9278 R29=0000000000000000 R30=00003FFF010DB0D8 R31=00003FFF7F5281D0
MLT stderr (0x00003FFEE5B11434 [<unknown>+0x0])
MLT stderr (0x00003FFF862E9840 [libj9vm29.so+0x169840])
MLT stderr NIP=00003FFEE5B11434 MSR=800000010280F033 ORIG_GPR3=00000000000081C8 CTR=00003FFF850BDC78
MLT stderr LINK=00003FFEE5B11428 XER=0000000000000000 CCR=0000000028004824 SOFTE=0000000000000001
MLT stderr TRAP=0000000000000300 DAR=0000000000000050 dsisr=0000000040000000 RESULT=0000000000000000
MLT stderr (0x00003FFF86199450 [libj9vm29.so+0x19450])
MLT stderr (0x00003FFF862343B0 [libj9vm29.so+0xb43b0])
MLT stderr (0x00003FFF860FD708 [libj9prt29.so+0x3d708])
MLT stderr (0x00003FFF8622F564 [libj9vm29.so+0xaf564])
MLT stderr (0x00003FFF860913E8 [libj9thr29.so+0x113e8])
MLT stderr (0x00003FFF87228040 [libpthread.so.0+0x8040])
MLT stderr clone+0x98 (0x00003FFF87143BB0 [libc.so.6+0x123bb0])
MLT stderr ---------------------------------------
...more
```

### PR DESCRIPTION
**Pull request by jdmpapin**
...when OpenJDK MethodHandles are enabled.

When the JIT processes a call to `invokeBasic()` on a known `MethodHandle`, `mh`, it loads `mh.form.vmentry` to find the `MemberName` that specifies the `LambdaForm`-generated method that implements `mh`. It then refines the call into a direct call to the implementing method. Typically this call will be inlined after refinement, but it may be generated as a call instead, e.g. due to `disableInlining`, or `dontInline={...}`, etc. Either way, the generated code may rely on the implementing method remaining loaded. For example, inlined code may need to resolve constant pool entries, load from pseudo-TOC entries that get released on unload, or OSR to continue execution in the interpreter.

The `MemberName` holds a reference to the generated class, so the implementing method will stay loaded as long as the `MemberName` is live. The references from the `MethodHandle` to the `LambdaForm` and then from there to the `MemberName` are usually enough to hold the `MemberName` live as long as the `MethodHandle`, so that the generated code works when the `invokeBasic()` call is reached at runtime. However, it is possible for some of these references to change after (or even during) compilation. In particular:

1. A `MethodHandle` may have its `LambdaForm` replaced with another semantically equivalent one.

2. A `LambdaForm` may be redundantly compiled to bytecode multiple times with no synchronization to prevent multiple distinct `MemberName` objects from being observed. The generated classes are only nominally different, so they're generally interchangeable.

If either of these situations arises after the JIT compiler has observed the `vmentry` for a particular `MethodHandle`, the update can sometimes make the original `vmentry` unreachable, and if it does, that can allow the generated bytecode to be unloaded even though it's still needed for a caller's JIT-compiled code.

This commit adds a field to `MethodHandle` called `jitVMEntryKeepAlive`. The first time the JIT compiler observes the `vmentry` corresponding to a given `MethodHandle`, it updates the field to point directly from the `MethodHandle` to the observed `MemberName` instance. Whenever the JIT subsequently looks for the `vmentry` of that `MethodHandle`, it will simply use the reference from `jitVMEntryKeepAlive`, ignoring `form.vmentry`. In this way, the compiler ensures that the `MemberName` it observes will live at least as long as the `MethodHandle`, and therefore so will the generated bytecode. The bytecode can still be unloaded once the `MethodHandle` is collected, but at that point any `invokeBasic()` call that was transformed for the handle had better already be unreachable.

Note that it's fine to miss the updates (1) and (2). For (1), the new `LambdaForm` is required to be equivalent to the old, and threads are allowed to keep running the old one indefinitely. In the case of (2), for a given `LambdaForm`, any generated class is as good as any other.

Fixes #13588

--------

There is one additional commit to prevent folding `MethodHandle.form` to a known object, just as a precaution.

### CODE REVIEW

**Comment by DanHeidinga** [comment]
@0xdaryl ping for 0.30

File: runtime/jcl/common/jclcinit.c [hunk 1]
```diff
@@ -634,6 +634,10 @@ initializeRequiredClasses(J9VMThread *vmThread, char* dllName)
 	if (0 != vmFuncs->addHiddenInstanceField(vm, "java/lang/invoke/MutableCallSite", "invalidationCookie", "J", &vm->mutableCallSiteInvalidationCookieOffset)) {
 		return 1;
 	}
+
+	if (0 != vmFuncs->addHiddenInstanceField(vm, "java/lang/invoke/MethodHandle", "jitVMEntryKeepAlive", "Ljava/lang/invoke/MemberName;", &vm->jitVMEntryKeepAliveOffset)) {
+		return 1;
+	}
```
Comment (babsingh) [inline runtime/jcl/common/jclcinit.c:640]: Minor formatting nitpick: indentation for lines 634-640 is different than the surrounding code.

**Review by babsingh — APPROVED** [review]
VM changes look good other than the formatting nitpick ... tentatively approving.

**Review by tajila — APPROVED** [review]

File: runtime/jcl/common/jclcinit.c (hunk 1 above)
Comment (jdmpapin) [reply runtime/jcl/common/jclcinit.c:640]: The indentation is actually exactly the same, which you should be able to see in an editor (e.g. with vim's `set list listchars=tab:>-`) by checking out this commit, copy-pasting from this diff, or downloading the raw contents of [jclcinit.c in this commit](https://github.com/jdmpapin/openj9/blob/4e9032ce962013c3c4598b002554a47d9a1b386e/runtime/jcl/common/jclcinit.c). Here's a [snippet that shows the indentation lining up visually](https://github.com/jdmpapin/openj9/blob/4e9032ce962013c3c4598b002554a47d9a1b386e/runtime/jcl/common/jclcinit.c#L625-L641) at least, though it doesn't visualize the whitespace

The apparent formatting problem looks to be some rendering quirk at the intersection of GitHub diffs, tabs, and Firefox

Comment (babsingh) [reply runtime/jcl/common/jclcinit.c:640]: Agreed, it's a browser issue. In Chrome, I do not notice the indentation issue.

**Comment by 0xdaryl** [comment]
Jenkins test sanity,extended all jdk17

**Comment by 0xdaryl** [comment]
@DanHeidinga @pshipton : While this is a problem that affects 0.30, it is a hard-to-reproduce corner case that is unlikely to occur in the field, which lowers its priority for 0.30 in my mind.  I think this can wait for JDK 18 (0.31).

It should be relatively safe for 0.30 though should anyone strongly assert that it be delivered in 0.30.

**Comment by pshipton** [comment]
Ok thanks. I'm moving the milestones to 0.32, as everything that isn't jdk18 specific is moved to this milestone.

**Comment by pshipton** [comment]
Note, just because something is in the 0.32 milestone plan doesn't mean it can't be delivered earlier and get into the 0.31 release.

**Comment by DanHeidinga** [comment]
> @DanHeidinga @pshipton : While this is a problem that affects 0.30, it is a hard-to-reproduce corner case that is unlikely to occur in the field, which lowers its priority for 0.30 in my mind. I think this can wait for JDK 18 (0.31).

Can you clarify why it's unlikely to occur in the field?  I've read the linked bug and it sounds like this is a race condition and therefore timing dependent.  What makes us believe that timing is unlikely?

**Comment by 0xdaryl** [comment]
Yes, this is a race condition.  That comment was based on our extraordinary attempts to catch the problem in order to even investigate it.  @jdmpapin, please offer an opinion on how likely such a race is to occur given your experience with it.

**Comment by jdmpapin** [comment]
Race conditions in general are rarely encountered unless intentionally provoked. We have only one test known to be affected by this bug, and during investigation the failure rate of that specific test was about 1/50.

Beyond that, this race condition was more difficult than I expected to intentionally provoke. Because the `LambdaForm` of a `MethodHandle` is compiled almost immediately after it is created (i.e. before we finish constructing the `MethodHandle`), the timing needed in the calling code to cause threads to race also generally prevented those threads from sharing the `LambdaForm`, since that is also created without synchronization and when the timing is close, each thread would create its own. So in addition to the delays I added to the JCL to exercise the directly related race condition, I also had to introduce a critical section to reproduce the issue.

Furthermore, even when the race condition occurs, the JIT compiler has to pick up the earlier `MemberName`. So we need something like this series of events:

1. Thread A creates the `LambdaForm` as part of creating a `MethodHandle`, and stores that `LambdaForm` into a cache for sharing. It goes on to start compiling to bytecode.
2. Before/during thread A's compilation of its `LambdaForm` to bytecode, but strictly after the cache store becomes visible, thread B is creating another `MethodHandle` that can share with that of thread A, and it gets its `LambdaForm` from the cache. Thread B starts compiling the same `LambdaForm` to bytecode because thread A hasn't finished yet.
3. Thread A finishes compiling to bytecode, sets the `MemberName` of the form, and finishes constructing its `MethodHandle`.
4. Before thread B is done compiling to bytecode, the JIT compiler starts compiling a method that is statically known to (directly or indirectly) call thread A's `MethodHandle` via `invokedynamic`, and reaches the point of determining the target method of the `MethodHandle`.
5. Finally, thread B finishes compiling the `LambdaForm` to bytecode and installs the new `MemberName`.

Compilation of a `LambdaForm` to bytecode is pretty quick, so steps 2 and 4 each have a very narrow time window in which to occur. Either one is unlikely by itself, but both have to happen (for a single `LambdaForm`) in order to encounter the bug.

As I mentioned in https://github.com/eclipse-openj9/openj9/issues/13588#issuecomment-991351351, it's also possible to encounter this bug due to a modification to `MethodHandle.form`. However, I don't believe such a failure has yet been observed outside of the test case that I attached to that comment. The most common reason for replacing the `LambdaForm` is customization, but a customized `LambdaForm` holds a reference to its uncustomized original, so customization can't cause this issue. During the investigation, I found only two other reasons for replacement:
1. A `DirectMethodHandle` that refers to a static member or constructor of a class that has not yet finished initialization will include a class initialization check. As soon as that check succeeds, it will replace the `LambdaForm` in the `DirectMethodHandle` with another one to remove the check.
2. When a `CountingWrapper` hits its count, it will replace its `LambdaForm` with one that no longer bothers counting, and that allows inlining of the target. The count for this is very small (30).

It's my expectation that such replacements are unlikely to occur after JIT compilation of methods inlining the handles.

Finally, I'll just mention that this bug is in the previous release, i.e. it's not a regression.

**Comment by 0xdaryl** [comment]
@jdmpapin : please see the x86-64 Linux crash in the CI build.  Not sure if it is related to this PR.

**Comment by jdmpapin** [comment]
That's #13577

**Comment by 0xdaryl** [comment]
> That's #13577

OK.  We should aim to make progress on that one for 0.32.  Merging this PR into master.

**Review by 0xdaryl — APPROVED** [review]
